### PR TITLE
Support Redis TLS/SSL Configuration

### LINKED ISSUE #8091: Support Redis TLS configuration
Similar to how some other connectors support TLS configuration, it would be great to have the same support for the Redis connector

More on securing Redis connection: https://redis.io/topics/encryption

### PR DESCRIPTION
**Pull request by maxstreese**
## Description

Starting with version 6, Redis allows connections to be TLS/SSL encrypted (see [here](https://redis.io/docs/management/security/encryption/)).

Currently, a Redis server that requires TLS/SSL encrypted connections is not supported by Trino, thereby limiting potential adoption of the tool.

This PR aims to implement the configuration property `redis.tls.enabled`, which when enabled, allows to connect to a TLS/SSL secured Redis server.

## Additional context and related issues

Fixes #8091

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Redis
* Add support for TLS connection. ({issue}`8091`)
```

### CODE REVIEW

File: docs/src/main/sphinx/connector/redis.rst [hunk 1]
```diff
@@ -66,6 +66,7 @@ Property name                           Description
 ``redis.database-index``                Redis database index
 ``redis.user``                          Redis server username
 ``redis.password``                      Redis server password
+``redis.use-ssl``                       Controls whether SSL is enabled or not
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/redis.rst:69]: Please change to `redis.tls.enabled`.

File: plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java [hunk 2]
```diff
@@ -220,13 +221,27 @@ public RedisConnectorConfig setRedisPassword(String redisPassword)
         return this;
     }
 
+    public boolean isUseSsl()
+    {
+        return useSsl;
+    }
+
+    @Config("redis.use-ssl")
+    @ConfigSecuritySensitive
```
Comment (ebyhr) [inline plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java:230]: Remove unnecessary `@ConfigSecuritySensitive` annotation. It's just a boolean flag.

File: plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java [hunk 3]
```diff
@@ -220,13 +221,27 @@ public RedisConnectorConfig setRedisPassword(String redisPassword)
...
+    @ConfigDescription("Whether SSL is enabled or not. Default is no")
```
Comment (ebyhr) [inline plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java:231]: No need to include the default value to the config description.

File: plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java [hunk 4]
```diff
@@ -220,13 +221,27 @@ public RedisConnectorConfig setRedisPassword(String redisPassword)
...
+    }
+
     public boolean isHideInternalColumns()
     {
         return hideInternalColumns;
     }
 
     @Config("redis.hide-internal-columns")
-    @ConfigDescription("Whether internal columns are shown in table metadata or not. Default is no")
+    @ConfigDescription("Whether internal columns are shown in table metadata or not. Default is yes")
```
Comment (ebyhr) [inline plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java:244]: Revert unrelated change.

**Review by ebyhr — COMMENTED** [review]
Is it possible to add an integration test?

File: docs/src/main/sphinx/connector/redis.rst [hunk 5]
```diff
@@ -180,14 +181,20 @@ The username for Redis server.
...
 
 This property is optional; the default is ``null``.
 
+``redis.use-ssl``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Starting with version 6, Redis supports TLS/SSL encrypted connections as an optional feature.
+In case the Redis server you want to connect to requires TLS/SSL, enabling this property will allow you to connect.
+
+This property is optional; the default is ``false``.
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/redis.rst:197]: @mosabua Do you want to review this docs?

File: plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java (hunk 4 above)
Comment (maxstreese) [reply plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java:244]: Understood. Should I create a dedicated commit for that though? The `no` is incorrect and also at odds with the `.rst` generated documentation (which is correctly stating that the default is `true`).

Comment (ebyhr) [reply plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java]: Unfortunately, this sentence is misleading. 

`hideInternalColumns = true` means that internal columns are **_not_** shown, so the existing **_no_** is correct. 
Also, bootstrap log prints the default config values. There's no need to mention the value in `@ConfigDescription`. If you still want to fix this line in this PR, yes please separate a commit.

File: plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java [hunk 6]
```diff
@@ -46,6 +46,7 @@
     private char redisKeyDelimiter = ':';
     private String redisUser;
     private String redisPassword;
+    private boolean useTls;
```
Comment (ebyhr) [inline plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java:49]: Rename to `tlsEnabled`.

**Comment by maxstreese** [comment]
> Please squash commits into one.

I'll do it once all outstanding issues with the PR are resolved.

> Is it possible to add an integration test?

I won't be able to do it today I think, but let me try tomorrow or Monday at the latest. I think it should be possible.

**Comment by ebyhr** [comment]
Please submit CLA if you haven't sent it.

**Comment by maxstreese** [comment]
> Please submit CLA if you haven't sent it.

Submitted yesterday (Friday, 28th of April) at 12:41 Europe/Berlin.

File: plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java (hunk 4 above)
Comment (maxstreese) [reply plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java]: Ah of course :man_facepalming:, my bad! I think it was just me not being smart then and this can be left untouched. I already reverted my "no to yes" change.

**Comment by maxstreese** [comment]
Just to give an update: I am still working on this. Trying to come up with an integration test turned out to be quite the endeavor for me. Apologies for taking my time.

**Comment by maxstreese** [comment]
Hi @ebyhr!

I finally managed to come around to implement this together with an integration test. I'll provide some pointers tomorrow.

**Comment by maxstreese** [comment]
Hi @ebyhr!

Please find below the promised points/pointers:

### 1. Additional TLS Configuration Properties

While trying to come up with an integration test, I stumbled upon [SslUtils](https://github.com/trinodb/trino/blob/8ea9b2040554c3536a06e755260ce9b2075e5d2a/lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/ssl/SslUtils.java#L42) and its usage in e.g. [CassandraClientModule](https://github.com/trinodb/trino/blob/8ea9b2040554c3536a06e755260ce9b2075e5d2a/plugin/trino-cassandra/src/main/java/io/trino/plugin/cassandra/CassandraClientModule.java#L60). I decided to integrate the setup for TLS connections from there into the Redis connector.

### 2. Changes To SSLUtils

For some reason, both the `keyStore` and the `trustStore` instances are created via `KeyStore.getInstance(KeyStore.getDefaultType())` calls. I am not a security expert by any means but while checking the `KeyStore` sources, I believe I found an alternative which should be more general (and is also required for the Redis TLS integration test) which is `KeyStore.getInstance(File file, char[] password)`. The documentation of this states (emphasis mine):

>    Returns a loaded keystore object **of the appropriate keystore type**.
>    First the keystore type is determined by probing the specified file.
>    Then a keystore object is instantiated and loaded using the data from that file.

Therefore - in my understanding - using this, rather than `getInstance` with the default type, you can handle different keystore types depending on the file. What this meant for the Redis TLS integration test is, that I did not have to convert the truststore file into JKS format which would have required additional steps either in the Redis image creation, or on the host (increasing the dependencies).

### 3. Failing Test In core/trino-main

It looks like on my final commit, there is a test failure in `core-trino-main`. I ran the test locally via `./mvnw test -pl core/trino-main -Dtest=TestDistinctLimitOperator` and things worked. I also could not see any connection to the changes I introduced, so my hope is that this test might just be flaky? I am happy to stand corrected here though :slightly_smiling_face:.

### 4. Alternative TLS Setup With Only `tls.enabled`

As I have outlined above, I believe with the additional TLS configuration properties, the Redis connector is more in line with the other connectors which already support TLS. I do however see, that this way of doing things, increases the number of code changes introduced with this PR quite significantly. If you therefore decide that this is too much, I believe there is an alternative which would only require the `tls.enabled` property and would still allow for the integration test to work. This alternative would work by setting system properties related to TLS/SSL. I have a working example for this [here](https://github.com/maxstreese/playground/blob/main/java/redis-tls-testcontainers/app/src/test/java/redistlstc/RedisWithTlsCustomViaProperties.java#L63).

### Closing Remarks

Apologies for causing so much work with a PR that I hoped would not cause too much trouble. Let me know if I can streamline this for you in any way.

File: docs/src/main/sphinx/connector/redis.rst [hunk 7]
```diff
@@ -66,6 +66,11 @@ Property name                           Description
 ``redis.database-index``                Redis database index
 ``redis.user``                          Redis server username
 ``redis.password``                      Redis server password
+``redis.tls.enabled``                   Whether TLS security is enabled, defaults to ``false``.
+``redis.tls.keystore-path``             Path to the key store.
+``redis.tls.keystore-password``         Password for the key store.
+``redis.tls.truststore-path``           Path to the trust store.
+``redis.tls.truststore-password``       Password for the trust store.
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/redis.rst:73]: Remove following dots for consistency with other properties.

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/ssl/SslUtils.java [hunk 8]
```diff
@@ -64,10 +62,7 @@ public static SSLContext createSSLContext(
             catch (IOException | GeneralSecurityException ignored) {
                 keyManagerPassword = keyStorePassword.map(String::toCharArray).orElse(null);
 
-                keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
-                try (InputStream in = new FileInputStream(keyStorePath.get())) {
-                    keyStore.load(in, keyManagerPassword);
-                }
+                keyStore = KeyStore.getInstance(keyStorePath.get(), keyManagerPassword);
```
Comment (ebyhr) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/ssl/SslUtils.java:65]: @Praveen2112 Do you want to review this change?

File: plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java [hunk 9]
```diff
@@ -246,6 +253,73 @@ public RedisConnectorConfig setKeyPrefixSchemaTable(boolean keyPrefixSchemaTable
...
+
+    @Config("redis.tls.enabled")
+    @ConfigDescription("Whether TLS security is enabled")
+    public RedisConnectorConfig setTlsEnabled(boolean tlsEnabled)
+    {
+        this.tlsEnabled = tlsEnabled;
+        return this;
+    }
+
+    public Optional<@FileExists File> getKeystorePath()
```
Comment (ebyhr) [inline plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisConnectorConfig.java:269]: Please consider extracting a config class like `MongoSslConfig`

File: plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisJedisManager.java [hunk 10]
```diff
@@ -105,12 +117,19 @@ public JedisPool getJedisPool(HostAddress host)
     private JedisPool createConsumer(HostAddress host)
     {
         log.info("Creating new JedisPool for %s", host);
+        var timeout = toIntExact(redisConnectTimeout.toMillis());
```
Comment (ebyhr) [inline plugin/trino-redis/src/main/java/io/trino/plugin/redis/RedisJedisManager.java:120]: Please avoid using `var`.

File: plugin/trino-redis/src/test/java/io/trino/plugin/redis/TestRedisConnectorConfig.java [hunk 11]
```diff
@@ -45,12 +48,20 @@ public void testDefaults()
...
+                .setTlsEnabled(false)
+                .setKeystorePath(null)
+                .setKeystorePassword(null)
+                .setTruststorePath(null)
+                .setTruststorePassword(null));
     }
 
     @Test
-    public void testExplicitPropertyMappings()
+    public void testExplicitPropertyMappings() throws IOException
```
Comment (ebyhr) [inline plugin/trino-redis/src/test/java/io/trino/plugin/redis/TestRedisConnectorConfig.java:60]: `throws IOException` on the new line. Throwing `Exception` is fine in test methods.

File: plugin/trino-redis/src/test/java/io/trino/plugin/redis/TestRedisTlsConnectorTest.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+import com.google.common.collect.ImmutableMap;
+import io.trino.plugin.redis.util.RedisSecurityFeature;
+import io.trino.plugin.redis.util.RedisServer;
+import io.trino.testing.QueryRunner;
+
+import static io.trino.plugin.redis.RedisQueryRunner.createRedisQueryRunner;
+import static io.trino.plugin.redis.util.RedisServer.LATEST_VERSION;
+
+public class TestRedisTlsConnectorTest
+        extends BaseRedisConnectorTest
```
Comment (ebyhr) [inline plugin/trino-redis/src/test/java/io/trino/plugin/redis/TestRedisTlsConnectorTest.java:25]: Running connector test to verify TLS configuration is overkill. Please consider introducing a smoke test.

File: plugin/trino-redis/src/test/java/io/trino/plugin/redis/TestRedisTlsConnectorTest.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+{
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        RedisServer redisServer = closeAfterClass(new RedisServer(LATEST_VERSION, RedisSecurityFeature.TLS));
+        var keyStorePath = redisServer.getContainerFilesDir().toPath().resolve("redis.p12").toString();
+        var trustStorePath = redisServer.getContainerFilesDir().toPath().resolve("ca.p12").toString();
```
Comment (ebyhr) [inline plugin/trino-redis/src/test/java/io/trino/plugin/redis/TestRedisTlsConnectorTest.java:33]: Avoid using `var`.

File: plugin/trino-redis/src/test/java/io/trino/plugin/redis/TestRedisTlsConnectorTest.java [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+        return createRedisQueryRunner(
+                redisServer,
+                ImmutableMap.of(),
+                ImmutableMap.of("redis.tls.enabled", "true", "redis.tls.keystore-path", keyStorePath, "redis.tls.keystore-password", "secret", "redis.tls.truststore-path", trustStorePath, "redis.tls.truststore-password", "secret"),
```
Comment (ebyhr) [inline plugin/trino-redis/src/test/java/io/trino/plugin/redis/TestRedisTlsConnectorTest.java:37]: This line is too long. Please wrap per entry.

File: plugin/trino-redis/src/test/java/io/trino/plugin/redis/util/RedisSecurityFeature.java [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.redis.util;
+
+public enum RedisSecurityFeature { NONE, USER_PASSWORD, TLS }
```
Comment (ebyhr) [inline plugin/trino-redis/src/test/java/io/trino/plugin/redis/util/RedisSecurityFeature.java:16]:
```suggestion
public enum RedisSecurityMode
{
    NONE,
    USER_PASSWORD,
    TLS,
}
```

File: plugin/trino-redis/src/test/java/io/trino/plugin/redis/util/RedisServer.java [hunk 16]
```diff
@@ -30,29 +45,77 @@
...
+
+                var trustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+                trustManagerFactory.init(trustStore);
+
+                var sslContext = SSLContext.getInstance("TLS");
+                sslContext.init(keyManagerFactory.getKeyManagers(), trustManagerFactory.getTrustManagers(), null);
+
+                jedisPool = new JedisPool(container.getHost(), container.getMappedPort(PORT), true, sslContext.getSocketFactory(), sslContext.getDefaultSSLParameters(), null);
+            }
+            default -> {
```
Comment (ebyhr) [inline plugin/trino-redis/src/test/java/io/trino/plugin/redis/util/RedisServer.java:106]: Change to `NONE`  and throw an exception in default.

File: plugin/trino-redis/src/test/java/io/trino/plugin/redis/util/RedisServer.java [hunk 17]
```diff
@@ -30,29 +45,77 @@
...
+                container = new GenericContainer<>("redis:" + version)
+                        .withCommand("redis-server", "--requirepass", PASSWORD)
+                        .withExposedPorts(PORT);
+                container.start();
+                jedisPool = new JedisPool(container.getHost(), container.getMappedPort(PORT), null, PASSWORD);
+                jedisPool.getResource().aclSetUser(USER, "on", ">" + PASSWORD, "~*:*", "+@all");
+            }
+            case TLS -> {
+                container = new GenericContainer(
+                        new ImageFromDockerfile("redistls", false)
```
Comment (ebyhr) [inline plugin/trino-redis/src/test/java/io/trino/plugin/redis/util/RedisServer.java:72]: `String version` is unused in case of TLS.

**Comment by ebyhr** [comment]
Could you rebase on master and address comments?

**Comment by maxstreese** [comment]
Hi @mosabua, apologies for the late reply. I think we can close this one for now. We have opted to solve the issue in our setup via a TLS tunnel for now.
